### DISCUSSION TITLE
GPS Time Sync Questions

### DISCUSSION THREAD

**Discussion by JStoschek** [gopro/labs#990] [Q&A] [answered]
Hello,

I have 3 GoPro Hero 10s. I am trying to get the exact time at which each frame from each GoPro was captured for computer vision tasks. I am looking at the GPS Time Synchronization feature and figured this might be helpful. I ran a test with one of my GoPros and got the below metadata when probing via ffmpeg.

 My question is what does the time in the timecode refer to? Is it current GPS time? or time since power on? Secondly, is there a better way to get all of the metadata from the MP4 other than ffmpeg prope? 
Any help would be greatly appreciated.

Metadata Snippet:
'tags': {'creation_time': '2024-07-16T14:47:05.000000Z',
                       'language': 'eng',
                       'handler_name': '\x0bGoPro H.265',
                       'vendor_id': '[0][0][0][0]',
                       'encoder': 'GoPro H.265 encoder',
                       'timecode': '04:15:32:29'}}

Full Metadata:
{'streams': [{'index': 0,
              'codec_name': 'hevc',
              'codec_long_name': 'H.265 / HEVC (High Efficiency Video Coding)',
              'profile': 'Main',
              'codec_type': 'video',
              'codec_tag_string': 'hvc1',
              'codec_tag': '0x31637668',
              'width': 3840,
              'height': 2160,
              'coded_width': 3840,
              'coded_height': 2160,
              'closed_captions': 0,
              'has_b_frames': 0,
              'sample_aspect_ratio': '1:1',
              'display_aspect_ratio': '16:9',
              'pix_fmt': 'yuvj420p',
              'level': 153,
              'color_range': 'pc',
              'color_space': 'bt709',
              'color_transfer': 'bt709',
              'color_primaries': 'bt709',
              'chroma_location': 'left',
              'refs': 1,
              'r_frame_rate': '30000/1001',
              'avg_frame_rate': '30000/1001',
              'time_base': '1/30000',
              'start_pts': 0,
              'start_time': '0.000000',
              'duration_ts': 354354,
              'duration': '11.811800',
              'bit_rate': '46831739',
              'nb_frames': '354',
              'disposition': {'default': 1,
                              'dub': 0,
                              'original': 0,
                              'comment': 0,
                              'lyrics': 0,
                              'karaoke': 0,
                              'forced': 0,
                              'hearing_impaired': 0,
                              'visual_impaired': 0,
                              'clean_effects': 0,
                              'attached_pic': 0,
                              'timed_thumbnails': 0},
              'tags': {'creation_time': '2024-07-16T14:47:05.000000Z',
                       'language': 'eng',
                       'handler_name': '\x0bGoPro H.265',
                       'vendor_id': '[0][0][0][0]',
                       'encoder': 'GoPro H.265 encoder',
                       'timecode': '04:15:32:29'}},
             {'index': 1,
              'codec_name': 'aac',
              'codec_long_name': 'AAC (Advanced Audio Coding)',
              'profile': 'LC',
              'codec_type': 'audio',
              'codec_tag_string': 'mp4a',
              'codec_tag': '0x6134706d',
              'sample_fmt': 'fltp',
              'sample_rate': '48000',
              'channels': 2,
              'channel_layout': 'stereo',
              'bits_per_sample': 0,
              'r_frame_rate': '0/0',
              'avg_frame_rate': '0/0',
              'time_base': '1/48000',
              'start_pts': 0,
              'start_time': '0.000000',
              'duration_ts': 566272,
              'duration': '11.797333',
              'bit_rate': '189591',
              'nb_frames': '553',
              'disposition': {'default': 1,
                              'dub': 0,
                              'original': 0,
                              'comment': 0,
                              'lyrics': 0,
                              'karaoke': 0,
                              'forced': 0,
                              'hearing_impaired': 0,
                              'visual_impaired': 0,
                              'clean_effects': 0,
                              'attached_pic': 0,
                              'timed_thumbnails': 0},
              'tags': {'creation_time': '2024-07-16T14:47:05.000000Z',
                       'language': 'eng',
                       'handler_name': '\x0bGoPro AAC  ',
                       'vendor_id': '[0][0][0][0]',
                       'timecode': '04:15:32:29'}},
             {'index': 2,
              'codec_type': 'data',
              'codec_tag_string': 'tmcd',
              'codec_tag': '0x64636d74',
              'r_frame_rate': '0/0',
              'avg_frame_rate': '30000/1001',
              'time_base': '1/30000',
              'start_pts': 0,
              'start_time': '0.000000',
              'duration_ts': 354354,
              'duration': '11.811800',
              'bit_rate': '2',
              'nb_frames': '1',
              'disposition': {'default': 1,
                              'dub': 0,
                              'original': 0,
                              'comment': 0,
                              'lyrics': 0,
                              'karaoke': 0,
                              'forced': 0,
                              'hearing_impaired': 0,
                              'visual_impaired': 0,
                              'clean_effects': 0,
                              'attached_pic': 0,
                              'timed_thumbnails': 0},
              'tags': {'creation_time': '2024-07-16T14:47:05.000000Z',
                       'language': 'eng',
                       'handler_name': '\x0bGoPro TCD  ',
                       'timecode': '04:15:32:29'}},
             {'index': 3,
              'codec_name': 'bin_data',
              'codec_long_name': 'binary data',
              'codec_type': 'data',
              'codec_tag_string': 'gpmd',
              'codec_tag': '0x646d7067',
              'r_frame_rate': '0/0',
              'avg_frame_rate': '0/0',
              'time_base': '1/1000',
              'start_pts': 0,
              'start_time': '0.000000',
              'duration_ts': 11812,
              'duration': '11.812000',
              'bit_rate': '53222',
              'nb_frames': '12',
              'disposition': {'default': 1,
                              'dub': 0,
                              'original': 0,
                              'comment': 0,
                              'lyrics': 0,
                              'karaoke': 0,
                              'forced': 0,
                              'hearing_impaired': 0,
                              'visual_impaired': 0,
                              'clean_effects': 0,
                              'attached_pic': 0,
                              'timed_thumbnails': 0},
              'tags': {'creation_time': '2024-07-16T14:47:05.000000Z',
                       'language': 'eng',
                       'handler_name': '\x0bGoPro MET  '}}],
 'format': {'filename': 'GX010093.MP4',
            'nb_streams': 4,
            'nb_programs': 0,
            'format_name': 'mov,mp4,m4a,3gp,3g2,mj2',
            'format_long_name': 'QuickTime / MOV',
            'start_time': '0.000000',
            'duration': '11.811800',
            'size': '69555764',
            'bit_rate': '47109340',
            'probe_score': 100,
            'tags': {'major_brand': 'mp41',
                     'minor_version': '538120216',
                     'compatible_brands': 'mp41',
                     'creation_time': '2024-07-16T14:47:05.000000Z',
                     'location': '+37.4***-122.0***/',
                     'location-eng': '+37.4***-122.0***/',
                     'firmware': 'H21.01.01.62.70'}}}

**Comment by dnewman-gpsw** [comment, marked as answer]
ffmpeg only get stock metadata, nothing custom, nothing frame based.  Use https://gopro.github.io/labs/control/metadata/ for more GoPro header metadata. But all the metadata you will need to code and extract the GPS5 track, GPS9 is HERO11 (better than HERO10, more designed for your needs.) Code access for all GPMF data: https://gopro.github.io/gpmf-parser/

**Reply by JStoschek** [reply]
Thank you, I think the Code access for all GPMF data will have what I want.

**Reply by xiemeilong** [reply]
How can I read the timecode track?  How many timecode in the track, one timecode  per frame or only one start/end timecode

**Reply by dnewman-gpsw** [reply]
Timecode track is in QuickTime format.  So dig into the Apple docs on MOV tracks and timecode.    
Or just use tools like FFMPEG that reads standard timecode.  It only stores the timecode for the first frame, which the most common for digital media.

**Reply by xiemeilong** [reply]
Thanks. Why is there only one frame in the track? `'nb_frames': '1',`

**Reply by xiemeilong** [reply]
I recorded the video after GPS Time Sync was enabled ， and I read the track using pyav， there is only one frame in the track. I found the metadata given by JStoschek also has one frame in the track.

**Reply by dnewman-gpsw** [reply]
Timecode is for the first frame, this is correct.

**Reply by Maxsteal91** [reply]
I have the same timecode for both the video track and the GPS track, but they are not aligned. Any suggestions on how to sync the frames with the GPS? I always get a slight offset of 3 to 4 seconds. I need an almost perfect alignment for a computer vision task.
